Bottom: b206b48c5db905b5b6305a8cafef496c88c697d3
Top:    f83bb9b0063c43d43bd04a79351244c75e37a773
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-01-07 03:18:32 +0000

Revert "hid/ps/ps.c: deal with zero sized circles even better."

This reverts commit 2eb76110abcc1bbed375b169808ef72c472cb604.

Instead of hardcoding a size, actually calculate the best fit.

This also makes the PostScript output of the circles test case
matching what's shown on the screen.


---

diff --git a/src/hid/ps/ps.c b/src/hid/ps/ps.c
index b74f7d1..771d212 100644
--- a/src/hid/ps/ps.c
+++ b/src/hid/ps/ps.c
@@ -1193,7 +1193,7 @@ ps_draw_arc (hidGC gc, Coord cx, Coord cy, Coord width, Coord height,
 	     Angle start_angle, Angle delta_angle)
 {
   Angle sa, ea;
-  double linewidth;
+  assert (width > 0);
 
   if (delta_angle > 0)
     {
@@ -1206,32 +1206,13 @@ ps_draw_arc (hidGC gc, Coord cx, Coord cy, Coord width, Coord height,
       ea = start_angle;
     }
 
-  use_gc (gc);
-
-  /* Other than pcb's screen renderer, PostScript (at least GhostScript)
-     internally limits linewidth to (diameter / 2), so no drawing of a dot with
-     a circle of zero diameter. Compensate for this by making diameter larger
-     and line width thinner.
-
-     Handling of this case currently has only circles in mind, no ellipses.
-     The regular case works with ellipses, too. */
-  if (width < global.linewidth)
-    {
-      Coord outer_radius;
-
-      outer_radius = width + global.linewidth / 2 + global.bloat;
-      width = height = outer_radius / 2;
-      linewidth = 2.0;
-    }
-  else
-    linewidth = (double) (global.linewidth + 2 * global.bloat) / (double) width;
-
-  /* The line of PostScript written here is a bit odd; linewidth isn't
-     absolute, but relative to circle diameter. This is neccessary for
-     ellipses, which are drawn by streching (transforming) a circle. */
-  pcb_fprintf (global.f, "%ma %ma %mi %mi %mi %mi %`g a\n",
-               sa, ea, -width, height, cx, cy, linewidth);
+  if (width <= 0)
+    width = MIL_TO_COORD (10);
 
+  use_gc (gc);
+  pcb_fprintf (global.f, "%ma %ma %mi %mi %mi %mi %g a\n",
+               sa, ea, -width, height, cx, cy,
+               (double) (global.linewidth + 2 * global.bloat) / (double) width);
 }
 
 static void
diff --git a/tests/golden/hid_ps1/circles.ps b/tests/golden/hid_ps1/circles.ps
index 3c7b56f..3fc8eb2 100644
--- a/tests/golden/hid_ps1/circles.ps
+++ b/tests/golden/hid_ps1/circles.ps
@@ -1,6 +1,6 @@
 %!PS-Adobe-3.0
 %%Title: circles.pcb
-%%CreationDate: Mon Sep 28 14:28:02 2015
+%%CreationDate: Mon Sep 28 14:02:45 2015
 %%Creator: PCB release: pcb 1.99z
 %%Version: (PCB pcb 1.99z) 0.0 0
 %%PageOrder: Ascend
@@ -17,7 +17,7 @@
 30 30 moveto (circles.pcb) show
 (1.) tocp
 (Table of Contents \(This Page\)) toc
-(Created on Mon Sep 28 14:28:02 2015
+(Created on Mon Sep 28 14:02:45 2015
 ) toc
 ( ) tocp
 (2.) tocp
@@ -257,9 +257,9 @@ stroke grestore
 0.57500 0.15000 0.57500 0.25000 t
 0.70000 0.07500 0.70000 0.32500 t
 0.05000 setlinewidth
-0 360 -0.01250 0.01250 0.30000 0.20000 2 a
+0 360 -0.01000 0.00000 0.30000 0.20000 5 a
 0.10000 setlinewidth
-0 360 -0.03750 0.03750 0.50000 0.20000 2 a
+0 360 -0.02500 0.02500 0.50000 0.20000 4 a
 0.00100 setlinewidth
 0 360 -0.10000 0.10000 0.70000 0.20000 0.01 a
 0.05000 setlinewidth
@@ -321,7 +321,7 @@ stroke grestore
 0.55000 0.55000 0.55000 0.45000 t
 0.70000 0.62500 0.70000 0.37500 t
 0.05000 setlinewidth
-0 360 -0.01250 0.01250 0.30000 0.50000 2 a
+0 360 -0.01000 0.00000 0.30000 0.50000 5 a
 0 360 -0.02500 0.02500 0.50000 0.50000 2 a
 0.00100 setlinewidth
 0 360 -0.10000 0.10000 0.70000 0.50000 0.01 a
@@ -382,9 +382,9 @@ stroke grestore
 0.57500 0.15000 0.57500 0.25000 t
 0.70000 0.07500 0.70000 0.32500 t
 0.05000 setlinewidth
-0 360 -0.01250 0.01250 0.30000 0.20000 2 a
+0 360 -0.01000 0.00000 0.30000 0.20000 5 a
 0.10000 setlinewidth
-0 360 -0.03750 0.03750 0.50000 0.20000 2 a
+0 360 -0.02500 0.02500 0.50000 0.20000 4 a
 0.00100 setlinewidth
 0 360 -0.10000 0.10000 0.70000 0.20000 0.01 a
 0.05000 setlinewidth
@@ -446,7 +446,7 @@ stroke grestore
 0.55000 0.55000 0.55000 0.45000 t
 0.70000 0.62500 0.70000 0.37500 t
 0.05000 setlinewidth
-0 360 -0.01250 0.01250 0.30000 0.50000 2 a
+0 360 -0.01000 0.00000 0.30000 0.50000 5 a
 0 360 -0.02500 0.02500 0.50000 0.50000 2 a
 0.00100 setlinewidth
 0 360 -0.10000 0.10000 0.70000 0.50000 0.01 a
@@ -1992,44 +1992,35 @@ stroke grestore
 2.78450 -0.22500 2.82200 -0.22500 t
 2.84000 -0.26250 2.84750 -0.26250 t
 2.84000 -0.24750 2.84750 -0.24750 t
+2.86550 -0.23250 2.87300 -0.22500 t
+2.86550 -0.27750 2.86550 -0.23250 t
 2.86550 -0.27750 2.87300 -0.28500 t
-2.87300 -0.28500 2.89550 -0.28500 t
-2.89550 -0.28500 2.90300 -0.27750 t
-2.90300 -0.27750 2.90300 -0.26250 t
-2.86550 -0.22500 2.90300 -0.26250 t
-2.86550 -0.22500 2.90300 -0.22500 t
-2.92100 -0.23250 2.92850 -0.22500 t
-2.92100 -0.24750 2.92100 -0.23250 t
-2.92100 -0.24750 2.92850 -0.25500 t
-2.92850 -0.25500 2.94350 -0.25500 t
-2.94350 -0.25500 2.95100 -0.24750 t
-2.95100 -0.24750 2.95100 -0.23250 t
-2.94350 -0.22500 2.95100 -0.23250 t
-2.92850 -0.22500 2.94350 -0.22500 t
-2.92100 -0.26250 2.92850 -0.25500 t
-2.92100 -0.27750 2.92100 -0.26250 t
-2.92100 -0.27750 2.92850 -0.28500 t
-2.92850 -0.28500 2.94350 -0.28500 t
+2.87300 -0.28500 2.88800 -0.28500 t
+2.88800 -0.28500 2.89550 -0.27750 t
+2.89550 -0.27750 2.89550 -0.23250 t
+2.88800 -0.22500 2.89550 -0.23250 t
+2.87300 -0.22500 2.88800 -0.22500 t
+2.86550 -0.24000 2.89550 -0.27000 t
+2.91350 -0.27750 2.92100 -0.28500 t
+2.92100 -0.28500 2.94350 -0.28500 t
 2.94350 -0.28500 2.95100 -0.27750 t
 2.95100 -0.27750 2.95100 -0.26250 t
-2.94350 -0.25500 2.95100 -0.26250 t
+2.91350 -0.22500 2.95100 -0.26250 t
+2.91350 -0.22500 2.95100 -0.22500 t
 2.96900 -0.26250 2.97650 -0.26250 t
 2.96900 -0.24750 2.97650 -0.24750 t
-2.99450 -0.23250 3.00200 -0.22500 t
-2.99450 -0.27750 2.99450 -0.23250 t
-2.99450 -0.27750 3.00200 -0.28500 t
-3.00200 -0.28500 3.01700 -0.28500 t
-3.01700 -0.28500 3.02450 -0.27750 t
-3.02450 -0.27750 3.02450 -0.23250 t
-3.01700 -0.22500 3.02450 -0.23250 t
-3.00200 -0.22500 3.01700 -0.22500 t
-2.99450 -0.24000 3.02450 -0.27000 t
-3.04250 -0.27750 3.05000 -0.28500 t
-3.05000 -0.28500 3.07250 -0.28500 t
-3.07250 -0.28500 3.08000 -0.27750 t
-3.08000 -0.27750 3.08000 -0.26250 t
-3.04250 -0.22500 3.08000 -0.26250 t
-3.04250 -0.22500 3.08000 -0.22500 t
+2.99450 -0.25500 3.02450 -0.28500 t
+2.99450 -0.25500 3.03200 -0.25500 t
+3.02450 -0.28500 3.02450 -0.22500 t
+3.05000 -0.28500 3.08000 -0.28500 t
+3.05000 -0.28500 3.05000 -0.25500 t
+3.05000 -0.25500 3.05750 -0.26250 t
+3.05750 -0.26250 3.07250 -0.26250 t
+3.07250 -0.26250 3.08000 -0.25500 t
+3.08000 -0.25500 3.08000 -0.23250 t
+3.07250 -0.22500 3.08000 -0.23250 t
+3.05750 -0.22500 3.07250 -0.22500 t
+3.05000 -0.23250 3.05750 -0.22500 t
 3.12500 -0.27750 3.13250 -0.28500 t
 3.13250 -0.28500 3.15500 -0.28500 t
 3.15500 -0.28500 3.16250 -0.27750 t
